### PR TITLE
Textmate scope selectors: add wildcard `*` and `-` negation

### LINKED ISSUE #52: TextMate scope selectors: scope exclusion is not implemented
_From @monk-time on September 4, 2017 14:51_

- VSCode Version: Code 1.15.1 (41abd21afdf7424c89319ee7cb0445cc6f376959, 2017-08-16T18:07:25.676Z)
- OS Version: Windows_NT x64 6.1.7601
- Extensions: 

Extension|Author (truncated)|Version
---|---|---
intellij-idea-keybindings|k--|0.2.13
selectline-statusbar|tom|0.0.2

Also the same result with today's VS Code Insiders build.

---

According to [the documentation](https://code.visualstudio.com/docs/extensions/themes-snippets-colorizers#_textmate-theme-rules) for TextMate scope selectors, VSCode supports the syntax for excluding matching scopes:
> `entity.name.method - source.java` matches all scopes that start with `entity.name.method` but not if a parent scope matches `source.java`

This functionality is used in at least one built-in theme that I could find:
https://github.com/Microsoft/vscode/blob/c00bdb74ee665cccfc5c4e41520893bb19ef61e5/extensions/theme-monokai/themes/monokai-color-theme.json#L316

But it seems that this syntax makes the selector with ` - ` in it invalid, and VSCode doesn't apply the given rule to anything at all.

Steps to Reproduce:

1. Put this rule for syntax highlighting in the settings. Confirm that comments turn yellow at least in some files.
```json
{
    "editor.tokenColorCustomizations": {
        "textMateRules": [
            {
                "scope": "comment",
                "settings": {
                    "foreground": "#FFFF00"
                }
            }
        ]
    }
}
```
2. Change the line with the scope selector to: 
`"scope": "comment - source.js",`
Expected result: comments that changed their color after step 1 remain yellow in all files except in `.js`.
Actual result: all comments reset to a color defined by the current theme.

Reproduces without extensions: Yes

_Copied from original issue: Microsoft/vscode#33802_

### LINKED ISSUE #160: Support wildcard `*` in scope selectors
Changes probably start in this file: https://github.com/microsoft/vscode-textmate/blob/v5.5.0/src/matcher.ts

This is a core part of Textmate specification: https://macromates.com/textmate/manual/references#scope-selector

### PR DESCRIPTION
**Pull request by SNDST00M**
Port of Atom's parser for Textmate scope selectors.
Resolves #52. Resolves #160.
PEG.js devDependencies: `pegjs@^0.10.0` `ts-pegjs@0.3.1`.

Script to regenerate the parser:
`npx pegjs --plugin ts-pegjs -o src/matcherParser.ts src/matcherParser.pegjs`

### CODE REVIEW

**Comment by SNDST00M** [comment]
Injection selectors were broken. I'm missing something here! 🤔

**Comment by SNDST00M** [comment]
@alexdima I'm confused about `grammar.ts`, could do with help!

**Comment by SNDST00M** [comment]
Found the problem, this is now ready to go! 🚀:shipit:

**Comment by SNDST00M** [comment]
<details>
<summary>Dev script for pegjs compiler inc. lockfile</summary>
```patch
diff --git a/package-lock.json b/package-lock.json
index 4691f43..b2bc326 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -1140,6 +1140,12 @@
       "integrity": "sha512-LDJzPVEEEPR+y48z93A0Ed0yXb8pAByGWo/k5YYdYgpY2/2EsOsksJrq7lOHxryrVOn1ejG6oAp8ahvOIQD8sw==",
       "dev": true
     },
+    "pegjs": {
+      "version": "0.10.0",
+      "resolved": "https://registry.npmjs.org/pegjs/-/pegjs-0.10.0.tgz",
+      "integrity": "sha1-z4uvrm7d/0tafvsYUmnqr0YQ3b0=",
+      "dev": true
+    },
     "pkg-dir": {
       "version": "4.2.0",
       "resolved": "https://registry.npmjs.org/pkg-dir/-/pkg-dir-4.2.0.tgz",
@@ -1430,6 +1436,15 @@
       "integrity": "sha1-DdTJ/6q8NXlgsbckEV1+Doai4fU=",
       "dev": true
     },
+    "ts-pegjs": {
+      "version": "0.3.1",
+      "resolved": "https://registry.npmjs.org/ts-pegjs/-/ts-pegjs-0.3.1.tgz",
+      "integrity": "sha512-4FqLW/HdZ4KvKHUpnaf7yeZM2L/1Y9cHlQ0Z59bVNABPbMEFQnKAYs/3p0zMvvCWkY3fnjMyr/F7wKzZ//vHIw==",
+      "dev": true,
+      "requires": {
+        "pegjs": "^0.10.0"
+      }
+    },
     "typescript": {
       "version": "4.5.4",
       "resolved": "https://registry.npmjs.org/typescript/-/typescript-4.5.4.tgz",
diff --git a/package.json b/package.json
index 59ae886..3cac83a 100644
--- a/package.json
+++ b/package.json
@@ -21,6 +21,7 @@
     "test": "tape -r ./out/tests/all.test.js",
     "benchmark": "node benchmark/benchmark.js",
     "inspect": "node out/tests/inspect.js",
+    "pegjs": "npx pegjs --plugin ts-pegjs -o src/matcherParser.ts src/matcherParser.pegjs",
     "tmconvert": "node scripts/tmconvert.js",
     "version": "npm run compile && npm run test && node scripts/release.js",
     "postversion": "git push && git push --tags",
@@ -29,7 +30,9 @@
   "devDependencies": {
     "@types/node": "^16.6.1",
     "@types/tape": "^4.13.2",
+    "pegjs": "^0.10.0",
     "tape": "^5.3.1",
+    "ts-pegjs": "^0.3.1",
     "typescript": "^4.4.3",
     "vscode-oniguruma": "^1.5.1",
     "webpack": "^5.50.0",
```

**Review by alexdima — APPROVED** [review]

File: src/matcherParser.pegjs [hunk 1]
```diff
@@ -1,12 +1,10 @@
...
- *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *  Copied from https://github.com/atom/first-mate/blob/master/src/scope-selector-parser.pegjs
  *
- *  Generated by PEG.js v0.10.0 (ts-pegjs plugin v0.3.1).
- *  https://pegjs.org/ (https://github.com/metadevpro/ts-pegjs)
+ *  Copyright (c) 2013 GitHub Inc. Licensed under the MIT License.
  *--------------------------------------------------------------------------------------------*/
 
-{ import matchers = require('./scopeSelectorMatchers'); }
+{ var matchers = require('./scope-selector-matchers'); }
```
[the remark this answers is not in the document]
Comment (alexdima) [reply src/matcherParser.pegjs]: :+1: Good catch! I've modified it to use the `customHeader` option of `tspegjs`

[the remark this answers is not in the document]
Comment (SNDST00M) [reply src/matcherParser.pegjs]: Ha nice! I like the CLI option much better

**Comment by tscpp** [comment]
> Adding support for * should be quite easily added to the current code.

I would really appreciate if this can be implemented. I need it to match the start tags for my injection grammar (i'm adding support for syntax highlighting of knockout attributes).

```
"injectionSelector": "L:meta.tag.structure.*.start",
```

**Comment by SNDST00M** [comment]
I understand the trepidation at adopting a parser especially a generated expression one, but this parsing is re-implemented many times in codebase and in core (with differences too). From outside, it seems like putting it in one place comes with strong benefit?

It's your call, I am using the port elsewhere so this didn't really consume my time

**Pull request closed without merging**